provider/openstack/provider.go

```diff
@@ -1261,7 +1261,53 @@ func (e *Environ) Instances(ids []instance.Id) ([]instance.Instance, error) {
 
 // AdoptResources is part of the Environ interface.
 func (e *Environ) AdoptResources(controllerUUID string, fromVersion version.Number) error {
-	return errors.NotImplementedf("AdoptResources")
+	var failed []string
+	controllerTag := map[string]string{tags.JujuController: controllerUUID}
+
+	instances, err := e.AllInstances()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cinder, err := e.cinderProvider()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// I'm not sure about this - it works at the moment because
```

Comment:
Oops - I meant to pull this out and replace it with a comment on github. Anyway, I'll pull it out once I know the answer.